### PR TITLE
feat: set max token limits for better user experience

### PR DESCRIPTION
**Pull request by BaseInfinity**
### Background
Every day we see people come into the discord explaining how they ran into a token limit because most people are still on a 3.0/3.5 key. The maintainer's built the app to assume 8000 token limit due to gpt4 key and atm this just causes friction because in reality, most people do not have a 4.0 key

To meet in the middle we will respect the user's token limit but compare it to the max token limit based off the smart/fast llvm model and set a limit 15% less to ensure we take into account the prompt and package

### Changes
Use the model's max_token_limits for both fast/smart llvm

### Test Plan
i updated failing tests

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

**Comment by rihp** [comment]
Shouldnt you also remove the variable from the .env template? I might be wrong

**Comment by BaseInfinity** [comment]
> Shouldnt you also remove the variable from the .env template? I might be wrong

I'm still letting people set custom limits in case they want to, I just make sure those custom limits are below a working threshold

**Comment by Boostrix** [comment]
FWIW, I actually exposed these limits via the command interface, so that the LLM can dynamically modify these once an error occurs.

The background being that I ran  into a bunch of errors where the hard-coded limits were not right, or were the hard-coded model was not working - exposing these details via the command interface (or even just via corresponding args) lets the AI/LLM play with different settings.

In my case it would figure out what to do over the course of the next 3-4 iterations - as long as you ensure to pass any error messages to the LLM and also pass in the command/arg to modify.

File: autogpt/config/config.py [hunk 1]
```diff
@@ -41,7 +41,24 @@ def __init__(self) -> None:
...
+
+        # It's better to set max limits and be explicit for better user experience IMO
+        model_limits = {
+            "gpt-4-32k": 32000,
+            "gpt-4": 8000,
+            "gpt-3.5-turbo": 4000,
+            "gpt-3.5-turbo-0301": 4000,
+        }
+
+        smart_token_limit = int(os.getenv("SMART_TOKEN_LIMIT", 4000))
```
Comment (k-boikov) [inline autogpt/config/config.py]: Why the default change from 8k to 4k?

Comment (BaseInfinity) [reply autogpt/config/config.py]: If for whatever reason it can't figure out what key it's on, 4000 seemed the safest. I don't mind changing it to 8000 and thinking more forward (assuming each model comes with an increase in token size?)

File: autogpt/config/config.py [hunk 2]
```diff
@@ -41,7 +41,24 @@ def __init__(self) -> None:
...
         self.fast_token_limit = int(os.getenv("FAST_TOKEN_LIMIT", 4000))
-        self.smart_token_limit = int(os.getenv("SMART_TOKEN_LIMIT", 8000))
+
+        # It's better to set max limits and be explicit for better user experience IMO
+        model_limits = {
+            "gpt-4-32k": 32000,
+            "gpt-4": 8000,
+            "gpt-3.5-turbo": 4000,
+            "gpt-3.5-turbo-0301": 4000,
+        }
```
Comment (ntindle) [inline autogpt/config/config.py]: Look through the files, I think these are already stored somewhere

**Comment by Boostrix** [comment]
in the light of the ongoing work to support multiple LLMs, it might make sense, to directly use a sub-dict named "openai" for these defaults ?

Also, like I said previously, exposing these via the command interface enables the llm to tinker with different settings if/when an error occurs - all that is needed is capturing the error via a system message that includes the commands/arguments that can be used to adapt these settings.

**Comment by BaseInfinity** [comment]
> in the light of the ongoing work to support multiple LLMs, it might make sense, to directly use a sub-dict named "openai" for these defaults ?
> 
> Also, like I said previously, exposing these via the command interface enables the llm to tinker with different settings if/when an error occurs - all that is needed is capturing the error via a system message that includes the commands/arguments that can be used to adapt these settings.

Correct me if I'm wrong but your way sounds like a better solution and is more future proof, if that's the case then I'm okay with closing this.

@ntindle if you want me to make the changes you suggested then I'll go ahead and do so but if this is going to be closed due to a better solution that's around the corner or close then I'm okay with letting this die lol

**Comment by Boostrix** [comment]
I would suggest to get this integrated with ntindle at this point, we will need for their feedback if they're supportive of the idea or not - with your current suggestion, at least you already got some feedback.

Also, some folks expressed concerns if hard-coded defaults are exposed to the LLM for tinkering, so what's probably going to happen is that someone is going to say we need to make this configurable, too

File: autogpt/config/config.py (hunk 2 above)
Comment (BaseInfinity) [reply autogpt/config/config.py]: https://github.com/Significant-Gravitas/Auto-GPT/pull/4128/commits/b3ba8581c157a23f1c0228f619945521dc38e116

kk I pull in from the constants from another file now

File: autogpt/config/config.py (hunk 1 above)
Comment (k-boikov) [reply autogpt/config/config.py]: Look at the line you deleted and replaced with this one, it used to be 8k. The smart llm by default is 4.0 so it should be 8k imo. https://help.openai.com/en/articles/7127966-what-is-the-difference-between-the-gpt-4-models

**Comment by BaseInfinity** [comment]
Ahh im failing the import tests, I'll fixie them this evening

File: autogpt/config/config.py (hunk 1 above)
Comment (BaseInfinity) [reply autogpt/config/config.py]: That's a fair point, Ill change it

File: autogpt/config/config.py (hunk 2 above)
Comment (BaseInfinity) [reply autogpt/config/config.py]: [2ad9647](https://github.com/Significant-Gravitas/Auto-GPT/pull/4128/commits/2ad96477c273db109eba59397873f47407f722f2)

File: autogpt/config/config.py (hunk 1 above)
Comment (BaseInfinity) [reply autogpt/config/config.py]: Changed back to 8k, thanks for the suggestion [2ad9647](https://github.com/Significant-Gravitas/Auto-GPT/pull/4128/commits/2ad96477c273db109eba59397873f47407f722f2)

**Comment by BaseInfinity** [comment]
Looks like tests are failing but not for my changes:
Screen Shot 2023-05-16 at 1 25 11 PM
/assets/1424113/3a8fc5a4-157a-4f05-9491-8c2d97cae7f8)

**Comment by BaseInfinity** [comment]
I need to update some of the tests to account for the new soft limits, I'll probably get to it before the end of the weekend

File: autogpt/config/config.py [hunk 3]
```diff
@@ -10,12 +10,16 @@
...
 
 
+
 class Config(metaclass=Singleton):
     """
     Configuration class to store the state of bools for different scripts access.
     """
 
     def __init__(self) -> None:
+        """Load here to avoid circular imports."""
```
Comment (ntindle) [inline autogpt/config/config.py]: do this after the """initialize the config class""" as it will replace the docstring for the function which is sub optimal

File: autogpt/config/config.py [hunk 4]
```diff
@@ -56,7 +60,24 @@ def __init__(self) -> None:
...
+            max_limit = smart_token_limit
+
+        if smart_token_limit > max_limit:
+            print(
+                f"Smart LLM model is {self.smart_llm_model}, setting max smart_token_limit to {max_limit}"
+            )
+            smart_token_limit = max_limit
+
+        self.smart_token_limit = smart_token_limit
+
```
Comment (ntindle) [inline autogpt/config/config.py]: Should probably set this limit for fast and smart not just smart

**Review by ntindle — CHANGES_REQUESTED** [review]

File: autogpt/config/config.py (hunk 3 above)
Comment (BaseInfinity) [reply autogpt/config/config.py]: [bd4637a](https://github.com/Significant-Gravitas/Auto-GPT/pull/4128/commits/bd4637a6277bd0b7b1568a2018d0ba5f1cb604f0)

File: autogpt/config/config.py (hunk 4 above)
Comment (BaseInfinity) [reply autogpt/config/config.py]: https://github.com/Significant-Gravitas/Auto-GPT/pull/4128/commits/d1546b7400e34a82ad8aab4cbbbc95129b4cdf3f

**Comment by ntindle** [comment]
Whoops sorry for updating, didn't realize you just pushed

**Comment by BaseInfinity** [comment]
@ntindle sorry for all the forced push commits, I couldn't run the same commands in CI as i could Locally and figured it would be faster to just git force push for meow versus attempting to run tests locally atm lol

File: autogpt/llm/providers/openai.py [hunk 5]
```diff
@@ -27,6 +33,7 @@
         prompt_token_cost=0.0004,
         completion_token_cost=0.0,
         max_tokens=8191,
+        soft_token_limit=int(8191 * SOFT_TOKEN_LIMIT),
         embedding_dimensions=1536,
```
Comment (BaseInfinity) [inline autogpt/llm/providers/openai.py]: i dont like feature creeping myself but should I do the same for embedding models lol? @ntindle

Comment (ntindle) [reply autogpt/llm/providers/openai.py]: Yes but not in this pr

File: tests/test_config.py [hunk 6]
```diff
@@ -9,6 +9,7 @@
 from autogpt.configurator import create_config
 
 
+# import
```
Comment (ntindle) [inline tests/test_config.py]: ?

File: autogpt/config/config.py [hunk 7]
```diff
@@ -55,8 +55,20 @@ def __init__(self) -> None:
         )
         self.fast_llm_model = os.getenv("FAST_LLM_MODEL", "gpt-3.5-turbo")
         self.smart_llm_model = os.getenv("SMART_LLM_MODEL", "gpt-4")
-        self.fast_token_limit = int(os.getenv("FAST_TOKEN_LIMIT", 4000))
-        self.smart_token_limit = int(os.getenv("SMART_TOKEN_LIMIT", 8000))
+
+        """ Load here to avoid circular imports. """
+        from autogpt.llm.providers.openai import OPEN_AI_CHAT_MODELS
+
+        self.OPEN_AI_CHAT_MODELS = OPEN_AI_CHAT_MODELS
```
Comment (ntindle) [inline autogpt/config/config.py]: Why is this needed?

**Comment by ntindle** [comment]
Don't forget to fill out the PR, I think our automaton uses it for change logs

File: tests/test_config.py (hunk 6 above)
Comment (BaseInfinity) [reply tests/test_config.py]: This is the result from running `isort` and/or `black` linters in order to get them passing

File: autogpt/config/config.py (hunk 7 above)
Comment (BaseInfinity) [reply autogpt/config/config.py]: since I decided to DRY things up and use a function, I needed a reference to the chat models. I could have passed it into the `get_max_token_limit` function but decided to assign it to the Config class instead

**Comment by BaseInfinity** [comment]
Ill fill out the description later today thanks for the review!

**Comment by BaseInfinity** [comment]
Fixed, thanks for the heads up @k-boikov

**Comment by Boostrix** [comment]
many people on discord still report that they find having to downgrade the hard-coded defaults to stay within the ctx window - so it would be good to get this updated/reviewed

**Comment by BaseInfinity** [comment]
@Boostrix sounds good, i fixed the conflicts to include soft limits on top of the recent memory config changes

**Review by Pwuts — COMMENTED** [review]
It would be nice to get this issue fixed, thanks for the PR!

Notes:
* I think we might as well remove the token limit from the config, and get the right token limit for the model in use from `autogpt/llm/providers` at runtime.
* I'm not convinced it's useful to set a global soft token limit, because the input/output ratio can vary a lot between cases where an LLM is used.
  Auto-GPT already calculates the maximum number of completion tokens based on the token limit and the number of tokens used by the prompt in `chat_with_ai()`, so just making sure the token limits are set properly should fix the issue. If not, something may be wrong with the token calculation in `chat_with_ai()`.

**Comment by BaseInfinity** [comment]
@Pwuts gotcha thanks for the info, i like those suggestions

**Comment by k-boikov** [comment]
image

**Comment by BaseInfinity** [comment]
I'll make suggested changes today and fix up the tests

**Comment by BaseInfinity** [comment]
@Pwuts I went ahead and removed both max token settings from the config file and now set them at runtime wherever it was used

**Review by ntindle — APPROVED** [review]

**Review by k-boikov — APPROVED** [review]

**Review by k-boikov — CHANGES_REQUESTED** [review]
What if people want to intentionally limit the tokens they spend? Sometimes that helps when we have a bug and we don't calculate the limits correctly. The only limit I think we should enforce if someone tries to play with their env and puts a limit higher than the model's max tokens.

**Comment by BaseInfinity** [comment]
> What if people want to intentionally limit the tokens they spend? Sometimes that helps when we have a bug and we don't calculate the limits correctly. The only limit I think we should enforce if someone tries to play with their env and puts a limit higher than the model's max tokens.

that's a fair point. I don't mind either solution we end up going with since it solves the major problem (mismanaged token limits).

At my day job I'd ping a project manager and ask for their 2 cents, is that @ntindle in this case lol?

**Comment by Pwuts** [comment]
> What if people want to intentionally limit the tokens they spend? Sometimes that helps when we have a bug and we don't calculate the limits correctly.

In such cases we should fix the bug rather than letting the user patch the bug with a setting. The possibility of bugs that may be alleviated by a setting is not a strong reason to keep that setting imo.

Aside from the above, are there any clear examples of cases when setting a lower token limit is desirable?

**Comment by BaseInfinity** [comment]
derp, i tried to sync up my master and accidentally wiped out my commits

**Comment by BaseInfinity** [comment]
I dont know what happened, my master somehow got divergent, I made a new pull request here since I can't open re-open this one now:
https://github.com/Significant-Gravitas/Auto-GPT/pull/4498
